keyring/backends/Windows.py

```diff
@@ -88,79 +301,152 @@ def priority(cls):
         If available, the preferred backend on Windows.
         """
         if missing_deps:
-            raise RuntimeError("Requires Windows and pywin32")
+            raise RuntimeError("Requires Windows and cffi")
         return 5
 
     @staticmethod
     def _compound_name(username, service):
-        return f'{username}@{service}'
+        return f"{username}@{service}"
+
+    def _discover_max_password_bytes(self):
+        def test_length(length):
+            try:
+                # NOTE relies on character "a" mapping to a single wchar_t (i.e. 2 bytes)
+                try:
+                    self._set_password("__probe_target__", "username", "a" * length)
+                except Exception as e:
+                    if not (isinstance(e, CredError) and e.winerror == 1783):
+                        log.exception(f"_set_password raised {e=}")
+                    raise
+                self._delete_password("__probe_target__")
+                return True
+            except CredError as e:
+                if e.winerror == 1783 and e.funcname == "CredWrite":
+                    return False
+                else:
+                    log.exception("unexpected CredError")
+                    return False
+            except Exception:
+                log.exception("unexpected Exception")
+
+        # first try to confirm observed max bytes of 2560
+        if test_length(2560) and not test_length(2561):
+            return 2560
+
+        # otherwise, use binary search
+        start, end = 1, MAX_PASSWORD_BYTES // 2
+        while start <= end:
+            mid = (start + end) // 2
+            if test_length(mid):
+                start = mid + 1
+            else:
+                end = mid - 1
+        return end
+
+    def __init__(self, *arg, **kw):
+        super().__init__(*arg, **kw)
+
+        # prevent sharding during self._discover_max_password_bytes
+        self._is_init = True
+        self._max_password_bytes = self._discover_max_password_bytes()
+        self._is_init = False
 
     def get_password(self, service, username):
         # first attempt to get the password under the service name
-        res = self._get_password(service)
-        if not res or res['UserName'] != username:
+        creds = self._get_password(service)
+        if not creds or creds["UserName"] != username:
             # It wasn't found so attempt to get it with the compound name
-            res = self._get_password(self._compound_name(username, service))
-        if not res:
+            creds = self._get_password(self._compound_name(username, service))
+        if not creds:
             return None
-        return res.value
+        password = creds.value
+        return password
```

Comment:
Once again, this change looks like an unnecessary stylistic change. It might be a useful change, but it distracts from the issue at hand.

Reply:
This is actually a meaningful change.  The previous code wan't checking if the the new username was different from the old username.  My change fixes #545.  I just now read your comment on #545, and  I agree that a unit test is in order.   

I really don't know the "convention" in this kind of thing.  My thinking was "well, I have the hood up, might as well fix this while I'm here." Perhaps this fix should be made in a separate PR?

Reply:
Agreed.